src/js/tech/setup-sourceset.js

```diff
@@ -42,42 +161,175 @@ const setupSourceset = function(tech) {
     srcDescriptor.enumerable = true;
   }
 
+  return srcDescriptor;
+};
+
+/**
+ * Patches browser internal functions so that we can tell syncronously
+ * if a `<source>` was appended to the media element. For some reason this
+ * causes a `sourceset` if the the media element is ready and has no source.
+ * This happens when:
+ * - The page has just loaded and the media element does not have a source.
+ * - The media element was emptied of all sources, then `load()` was called.
+ *
+ * It does this by patching the following functions/properties when they are supported:
+ *
+ * - `append()` - can be used to add a `<source>` element to the media element
+ * - `appendChild()` - can be used to add a `<source>` element to the media element
+ * - `insertAdjacentHTML()` -  can be used to add a `<source>` element to the media element
+ * - `innerHTML` -  can be used to add a `<source>` element to the media element
+ *
+ * @param {Html5} tech
+ *        The tech object that sourceset is being setup on.
+ */
+const firstSourceWatch = function(tech) {
+  const el = tech.el();
+
+  if (el.firstSourceWatch_) {
+    return;
+  }
+
+  el.firstSourceWatch_ = true;
+  const oldAppend = el.append;
+  const oldAppendChild = el.appendChild;
+  const oldInsertAdjacentHTML = el.insertAdjacentHTML;
+  const innerDescriptor = getInnerHTMLDescriptor(el);
+
+  el.appendChild = function() {
+    const retval = oldAppendChild.apply(el, arguments);
+
+    sourcesetLoad(tech);
+
+    return retval;
+  };
+
+  if (oldAppend) {
+    el.append = function() {
+      const retval = oldAppend.apply(el, arguments);
+
+      sourcesetLoad(tech);
+
+      return retval;
+    };
+  }
+
+  if (oldInsertAdjacentHTML) {
+    el.insertAdjacentHTML = function() {
+      const retval = oldInsertAdjacentHTML.apply(el, arguments);
+
+      sourcesetLoad(tech);
+
+      return retval;
+    };
+  }
+
+  Object.defineProperty(el, 'innerHTML', {
+    get: innerDescriptor.get.bind(el),
+    set(v) {
+      const retval = innerDescriptor.set.call(el, v);
+
+      sourcesetLoad(tech);
+
+      return retval;
+    },
+    configurable: true,
+    enumerable: innerDescriptor.enumerable
+  });
+
+  // on the first sourceset, we need to revert
+  // our changes
+  tech.one('sourceset', (e) => {
```

Comment:
and because all of the things above trigger sourceset, we'd reset this once the first source was set?

Reply:
yes